### PR TITLE
NDEV-2470 Evict old transactions from the mempool

### PR DESCRIPTION
**Pull request by alfiedotwtf**
Draft PR of NDEV-2470 to solicit feedback.

### CODE REVIEW

File: proxy/mempool/mempool.py [hunk 1]
```diff
@@ -523,6 +524,13 @@ def on_gas_price(self, gas_price: MPGasPriceResult) -> None:
     def _create_kick_tx_schedule_task(self) -> None:
         asyncio.get_event_loop().create_task(self._kick_tx_schedule())
 
+    async def _process_tx_eviction_loop(self) -> None:
+        while True:
+            for tx_schedule in self._tx_schedule_dict.values():
+                tx_schedule._drop_old_txs()
+
+            await asyncio.sleep(self._config.mempool_eviction_time_sec)
```
Comment (afalaleev) [inline proxy/mempool/mempool.py]: If the eviction time is 3 hours, I can send a tx that can live in the mempool <=6 hours.

- loop is started
- loop is finished
- mempool receives a tx
- loop is started
- loop skips tx, because the heartbeat is less than 3 hours
- loop is finished

File: proxy/mempool/mempool_schedule.py [hunk 2]
```diff
@@ -390,6 +432,8 @@ def add_tx(self, tx: MPTxRequest) -> MPTxSendResult:
                 )
                 self._drop_tx_from_sender_pool(sender_pool, old_tx)
 
+        tx.update_heartbeat()
```
Comment (afalaleev) [inline proxy/mempool/mempool_schedule.py]: The heartbeat is already set in the ctor of MPTxRequest

**Review by afalaleev — CHANGES_REQUESTED** [review]
Tests fail because they create a list of test MPTxRequests.

But it isn't the main flaw.
In the issue NDEV-2470 you described the algorithm for removing ALL transactions if the SENDER didn't do anything last 3 hours.
The implementation removes ONLY transactions with an age of more than 3 hours.

So, we have two different results.
Case 1.
Described algorithm:
-- User (tx counter = 5) sends the 1st transaction (nonce = 7)
-- After 2:59 User sends the 2nd transaction (nonce = 6)
-- After 3:01 User sends the 3rd transaction (nonce = 5)
-- Mempool executes 3 transactions (nonce 5, 6, 7)

Implemented algorithm:
-- User (tx counter = 5) sends the 1st transaction (nonce = 7)
-- After 2:59 User sends the 2nd transaction (nonce = 6)
-- Mempool removes the 1st transaction (nonce = 7)
-- After 3:01 User sends the 3rd transaction (nonce = 5)
-- Mempool executes 2 transactions (nonce 5, 6)

Case 2
Described algorithm:
-- User (tx counter = 5) sends the 1st transaction (nonce = 7)
-- After 2:59 User sends the 2nd transaction (nonce = 6)
-- After 6 hours Mempool removes 2 transactions (nonce 6, 7)

Implemented algorithm:
-- User (tx counter = 5) sends the 1st transaction (nonce = 7)
-- After 2:59 User sends the second transaction (nonce = 6)
-- After 3 hours Mempool removes 1 transaction (nonce 7)
-- After 6 hours Mempool removes 1 transaction (nonce 6)

**Comment by alfiedotwtf** [comment]
> But it isn't the main flaw. In the issue NDEV-2470 you described the algorithm for removing ALL transactions if the SENDER didn't do anything last 3 hours. The implementation removes ONLY transactions with an age of more than 3 hours.

I thought a lot about this, and I think the current implementation is better than what was originally designed...

The way it is currently implemented gives us:

  - Transactions are good if they CAN run before timeout
  - Transactions are bad if they CANNOT run before timeout
  - Reward by leaving good transactions at timeout
  - Punish by evicting bad transactions at timeout
 
In other word, here we only evict bad transactions.

Compare this to the original design:

  - A sender is good if they have touched their queue before timeout
  - A sender is bad if they have not touched their queue before timeout
  - Reward by leaving good transactions at timeout
  - Reward by leaving BAD transactions at timeout

In other words, a malicious sender can fill the mempool by simply "touching bad transactions in their queue".

I'm happy to change it back to how I originally had the code, but after looking at the issue of "malicious touchers", I went with the current implementation instead.

Hope that makes sense.

File: proxy/mempool/mempool_schedule.py (hunk 2 above)
Comment (alfiedotwtf) [reply proxy/mempool/mempool_schedule.py]: For the "new" design, we should always update the heartbeat on either:

  - insertion
  - fee replacement

This update here catches both scenarios. If we do not do this here, then the only time the heartbeat is set is during construction.

File: proxy/mempool/mempool.py (hunk 1 above)
Comment (alfiedotwtf) [reply proxy/mempool/mempool.py]: In `_drop_old_txs(self)` we have:

    tx = self._tx_dict.peek_heartbeat_tx()

Here, `peek_heartbeat_tx()` is a sorted queue, so should return older transactions before newer transactions.

So in step 3 of your example above, when the mempool receives a new tx, yes it will be skipped because this individual transaction has a heartbeat is less than 3 hours. In other words, we should not be evicting new transactions, only old ones.

**Comment by alfiedotwtf** [comment]
After talking with @afalaleev, it was decided to go back to the original design - put the heartbeats on the sender rather than individual transactions.

**Comment by alfiedotwtf** [comment]
Closing this in favour of https://github.com/neonlabsorg/proxy-model.py/pull/249
